ath79: add support for TP-Link TL-WR841HP v2

saif-ibne-ali opened the pull request:
This is a patch for **TP-Link TL-WR841HP v2**

**Specifications:**
- AR9344 SoC, 8 MB nor flash, 64 MB DDR2 RAM
- 2x2 9dBi antenna, wifi 2.4Ghz 300Mbps
- 4x Ethernet LAN 10/100, 1x Ethernet WAN 10/100
- 1x WAN, 4x LAN, Wifi, PWR, WPS, SYSTEM Leds
- Reset/WPS button
- Serial UART at J4 onboard: 3.3v GND RX TX, 1152008N1

**MAC addresses as verified by OEM firmware:**

Vendor-----OpenWrt---Address
LAN--------eth0--------label
WAN-------eth1--------label (+1)
WLAN------phy0-------label

The label MAC address was found in u-boot 0x1fc00.

**Installation:**
To install openwrt,
 - set the device's SSID to each of the following lines,
   making sure to include the backticks.
 - set the ssid and click save between each line.
```
`echo "httpd -k"> /tmp/s`
`echo "sleep 10">> /tmp/s`
`echo "httpd -r&">> /tmp/s`
`echo "sleep 10">> /tmp/s`
`echo "httpd -k">> /tmp/s`
`echo "sleep 10">> /tmp/s`
`echo "httpd -f">> /tmp/s`
`sh /tmp/s`
```
- Now, wait 60 sec.
- After the reboot sequence, the router may have fallen back to
  its default IP address with the default credentials (admin:admin).
- Log in to the web interface and go the the firmware upload page.
  Select "openwrt-ath79-generic-tplink_tl-wr841hp-v2-squashfs-factory.bin"
  and you're done : the system now accepts the openwrt.

**Forum support topic:**
https://forum.openwrt.org/t/support-for-tplink-tl-wr841hp-v2/69445/

Signed-off-by: Saiful Islam <si87868@gmail.com>

adschm commented:
Hi, Signed-off-by needs to be full real name.

adschm commented on target/linux/ath79/dts/ar9344_tplink_tl-wr841hp-v2.dts:
```diff
@@ -0,0 +1,157 @@
+		led-running = &led_system;
+		led-upgrade = &led_system;
+		label-mac-device = &wmac;
+	};
```
Missing empty line between nodes.

Human-OS replied:
missing line added.

adschm commented on target/linux/ath79/dts/ar9344_tplink_tl-wr841hp-v2.dts:
```diff
@@ -0,0 +1,157 @@
+		led-upgrade = &led_system;
+		label-mac-device = &wmac;
+	};
+	leds: leds {
```
"leds:" DT is not used, remove it.

Human-OS replied:
it is removed. thanks

adschm commented on target/linux/ath79/dts/ar9344_tplink_tl-wr841hp-v2.dts:
```diff
@@ -0,0 +1,157 @@
+	leds: leds {
+		compatible = "gpio-leds";
+
+		wlan {
```
Please sort leds by something, either name or GPIO number.

Human-OS replied:
Sorting is done with respect to GPIO number.

adschm commented on target/linux/ath79/dts/ar9344_tplink_tl-wr841hp-v2.dts:
```diff
@@ -0,0 +1,157 @@
+			label = "green:lan4";
+			gpios = <&gpio 15 GPIO_ACTIVE_LOW>;
+		};
+	};
```
Missing empty line.

Human-OS replied:
line has added before "keys :" node.

adschm commented on target/linux/ath79/dts/ar9344_tplink_tl-wr841hp-v2.dts:
```diff
@@ -0,0 +1,157 @@
+			gpios = <&gpio 4 GPIO_ACTIVE_LOW>;
+			debounce-interval = <60>;
+		};
+
```
remove empty line

Human-OS replied:
Done.

adschm commented on target/linux/ath79/generic/base-files/etc/board.d/02_network:
```diff
@@ -411,6 +411,7 @@ ath79_setup_interfaces()
 			"0@eth0" "2:lan:3" "3:lan:2" "4:lan:1" "5:lan:4"
 		;;
 	tplink,tl-wr842n-v2|\
+	tplink,tl-wr841hp-v2|\
```
Alphabetic sorting.

Human-OS replied:
sorting is done. Thanks

Human-OS replied:
Every requested changes has done. Please Check if more changes are needed. Thank you for your time.

adschm reviewed (changes requested):
Nothing serious on a first look.

Please work lines in the commit message to 74 chars. max.

Please also add a MAC address overview table:
https://openwrt.org/docs/guide-developer/device-support-policies#mac_addresses

teqlee commented on target/linux/ath79/dts/ar9344_tplink_tl-wr841hp-v2.dts:
```diff
@@ -0,0 +1,157 @@
+			default-state = "on";
+		};
+
+		qss {
```
I think should rename qss to wps for the same as the user manual. And what about power led ?

teqlee commented on target/linux/ath79/dts/ar9344_tplink_tl-wr841hp-v2.dts:
```diff
@@ -0,0 +1,157 @@
+			default-state = "on";
+		};
+
+		qss {
+			label = "green:qss";
```
I think should rename qss to wps for the same as the manual. And what about power led ?

Human-OS replied:
qss has changed to wps. Power led do not have any gpio pin, its directly related to power source.

teqlee commented on target/linux/ath79/generic/base-files/etc/board.d/02_network:
```diff
@@ -411,6 +411,7 @@ ath79_setup_interfaces()
+	tplink,tl-wr841hp-v2|\
 	tplink,tl-wr941hp-v1)
 		ucidef_set_interface_wan "eth1"
 		ucidef_add_switch "switch0" \
```
are you sure the port order is correct ?

Human-OS replied:
Port order is confirmed. it has taken 2 days to figure out and i have found a exact match of ports with tl-wr842n-v2.

Human-OS commented:
> Hi, Signed-off-by needs to be full real name.

DCO force me to use github username. Then how i will add real user name?

csharper2005 commented:
> > Hi, Signed-off-by needs to be full real name.
> 
> DCO force me to use github username. Then how i will add real user name?

Put you real name in your commit, I.e.
Signed-off-by: Real Name <si87868@gmail.com>

Human-OS commented:
> > > Hi, Signed-off-by needs to be full real name.
> > 
> > 
> > DCO force me to use github username. Then how i will add real user name?
> 
> Put you real name in your commit, I.e.
> Signed-off-by: Real Name [si87868@gmail.com](mailto:si87868@gmail.com)

Sign-off using real name is done. Thank you @adschm & @csharper2005 .

Human-OS commented:
> Nothing serious on a first look.
> 
> Please work lines in the commit message to 74 chars. max.
> 
> Please also add a MAC address overview table:
> https://openwrt.org/docs/guide-developer/device-support-policies#mac_addresses

Hey @adschm . All the changes requested by you, has been completed. Can you check again, Sir?

Human-OS commented:
**note1 from ar9344.dtsi**
`&wmac {
	interrupt-parent = <&intc2>;
	interrupts = <0>;
};`

**note2 from ar9341.dtsi**
`&wmac {
	interrupt-parent = <&cpuintc>;
	interrupts = <2>;
};`

wr841hpv2 is only target in ar9344 without 5ghz.
But When i build openwrt set target as ar9344 everything is work except huge wireless performance & range degradation.
On the other hand  when openwrt is built set target as ar9341 , Wireless works as expected. But all gpio button or led not work.
Also a different behavior found in  atheros wireless driver in two different builds
_in ar9341 build:
ieee80211 phy0: Atheros AR9340 Rev:2 mem=0xb8100000, irq=2
in ar9344 build:
ieee80211 phy0: Atheros AR9340 Rev:2 mem=0xb8100000, irq=12_

teqlee commented:
> 
> i don't have any idea to solve this problem, So, need some help to fix this problem.
irq depend on difference revision cpu, i think you missing gpio to control tx power amplifier circuit.

Human-OS commented:
> irq depend on difference revision cpu, i think you missing gpio to control tx power amplifier circuit.

then, how will i code this gpio in dts file?

teqlee commented:
you have to figure out for yourself which gpios control the txpower, as well as how you find out the other gpios.
based on your stock bootlog, I'm guessing that those gpio's are 18 & 19.
`++++++++++++++gpio_xlna_enable@543:xlna_chain0=18,enable=1:
++++++++++++++gpio_xlna_enable@554:xlna_chain1=19,enable=1:`

Human-OS commented:
> you have to figure out for yourself which gpios control the txpower, as well as how you find out the other gpios.
> based on your stock bootlog, I'm guessing that those gpio's are 18 & 19.
> `++++++++++++++gpio_xlna_enable@543:xlna_chain0=18,enable=1: ++++++++++++++gpio_xlna_enable@554:xlna_chain1=19,enable=1:`
> 
> please refer to ar9344_tplink_cpe.dtsi example:
> 
> > &gpio {
> > gpio_ext_lna0 {
> > gpio-hog;
> > gpios = <18 GPIO_ACTIVE_HIGH>;
> > output-high;
> > line-name = "tp-link:ext:lna0";
> > };

Yeah ! Now wireless working perfectly. You saved my times. Thanks a lot @teqlee .

Human-OS commented:
Already 1 month has passed, but i do not know why this pr is still blocked from merging. What criterias are missing?

csharper2005 commented:
> Already 1 month has passed, but i do not know why this pr is still blocked from merging. What criterias are missing?

Never mind, this is a common practice. It's open source, no any SLAs and strict criteria are here.

adschm commented on target/linux/ath79/dts/ar9344_tplink_tl-wr841hp-v2.dts:
```diff
@@ -0,0 +1,172 @@
+&eth1 {
+	status = "okay";
+
+	phy-handle = <&swphy4>;
```
I've never seen a setup like this where phy-handle is on both ethX.

Is it possible that you can simply drop the phy-handle for eth1?

Human-OS replied:
> I've never seen a setup like this where phy-handle is on both ethX.
> 
> Is it possible that you can simply drop the phy-handle for eth1?

I have tested dropping phy-handle for eth1. Ports are working fine as before. Now, it is also taking less time to boot the system. Thank you a lot @adschm

saif-ibne-ali committed:
ath79: add support for TP-Link TL-WR841HP v2

Specifications:
- AR9344 SoC, 8 MB nor flash, 64 MB DDR2 RAM
- 2x2 9dBi antenna, wifi 2.4Ghz 300Mbps
- 4x Ethernet LAN 10/100, 1x Ethernet WAN 10/100
- 1x WAN, 4x LAN, Wifi, PWR, WPS, SYSTEM Leds
- Reset/WPS button
- Serial UART at J4 onboard: 3.3v GND RX TX, 1152008N1

MAC addresses as verified by OEM firmware:

vendor   OpenWrt   address
LAN      eth0      label
WAN      eth1      label + 1
WLAN     phy0      label

The label MAC address was found in u-boot 0x1fc00.

Installation:
To install openwrt,
- set the device's SSID to each of the following lines,
  making sure to include the backticks.
- set the ssid and click save between each line.

`echo "httpd -k"> /tmp/s`
`echo "sleep 10">> /tmp/s`
`echo "httpd -r&">> /tmp/s`
`echo "sleep 10">> /tmp/s`
`echo "httpd -k">> /tmp/s`
`echo "sleep 10">> /tmp/s`
`echo "httpd -f">> /tmp/s`
`sh /tmp/s`

- Now, wait 60 sec.
- After the reboot sequence, the router may have fallen back to
  its default IP address with the default credentials (admin:admin).
- Log in to the web interface and go the the firmware upload page.
  Select "openwrt-ath79-generic-tplink_tl-wr841hp-v2-squashfs-factory.bin"
  and you're done : the system now accepts the openwrt.

Forum support topic:
https://forum.openwrt.org/t/support-for-tplink-tl-wr841hp-v2/69445/

Signed-off-by: Saiful Islam <si87868@gmail.com>

adschm commented:
Could you provide a recent bootlog (or dmesg output)?

Human-OS commented:
> Could you provide a recent bootlog (or dmesg output)?

```
root@OpenWrt:~# dmesg
[    0.000000] Linux version 5.4.137 (human@HP-Notebook) (gcc version 8.4.0 (OpenWrt GCC 8.4.0 r16256-2d5ee43dc6)) #0 Sat Jul 31 17:21:01 2021
[    0.000000] printk: bootconsole [early0] enabled
[    0.000000] CPU0 revision is: 0001974c (MIPS 74Kc)
[    0.000000] MIPS: machine is TP-Link TL-WR841HP v2
[    0.000000] SoC: Atheros AR9344 rev 2
[    0.000000] Initrd not found or empty - disabling initrd
[    0.000000] Primary instruction cache 64kB, VIPT, 4-way, linesize 32 bytes.
[    0.000000] Primary data cache 32kB, 4-way, VIPT, cache aliases, linesize 32 bytes
[    0.000000] Zone ranges:
[    0.000000]   Normal   [mem 0x0000000000000000-0x0000000003ffffff]
[    0.000000] Movable zone start for each node
[    0.000000] Early memory node ranges
[    0.000000]   node   0: [mem 0x0000000000000000-0x0000000003ffffff]
[    0.000000] Initmem setup node 0 [mem 0x0000000000000000-0x0000000003ffffff]
[    0.000000] On node 0 totalpages: 16384
[    0.000000]   Normal zone: 144 pages used for memmap
[    0.000000]   Normal zone: 0 pages reserved
[    0.000000]   Normal zone: 16384 pages, LIFO batch:3
[    0.000000] pcpu-alloc: s0 r0 d32768 u32768 alloc=1*32768
[    0.000000] pcpu-alloc: [0] 0 
[    0.000000] Built 1 zonelists, mobility grouping on.  Total pages: 16240
[    0.000000] Kernel command line: console=ttyS0,115200 rootfstype=squashfs,jffs2
[    0.000000] Dentry cache hash table entries: 8192 (order: 3, 32768 bytes, linear)
[    0.000000] Inode-cache hash table entries: 4096 (order: 2, 16384 bytes, linear)
[    0.000000] Writing ErrCtl register=00000000
[    0.000000] Readback ErrCtl register=00000000
[    0.000000] mem auto-init: stack:off, heap alloc:off, heap free:off
[    0.000000] Memory: 57152K/65536K available (4969K kernel code, 189K rwdata, 1136K rodata, 1184K init, 196K bss, 8384K reserved, 0K cma-reserved)
[    0.000000] SLUB: HWalign=32, Order=0-3, MinObjects=0, CPUs=1, Nodes=1
[    0.000000] NR_IRQS: 51
[    0.000000] random: get_random_bytes called from start_kernel+0x358/0x54c with crng_init=0
[    0.000000] CPU clock: 560.000 MHz
[    0.000000] clocksource: MIPS: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 6825930166 ns
[    0.000009] sched_clock: 32 bits at 280MHz, resolution 3ns, wraps every 7669584382ns
[    0.008315] Calibrating delay loop... 278.78 BogoMIPS (lpj=557568)
[    0.050863] pid_max: default: 32768 minimum: 301
[    0.055965] Mount-cache hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.063765] Mountpoint-cache hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.077270] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 7645041785100000 ns
[    0.087702] futex hash table entries: 256 (order: -1, 3072 bytes, linear)
[    0.095089] pinctrl core: initialized pinctrl subsystem
[    0.101849] NET: Registered protocol family 16
[    0.114469] GPIO line 18 (tp-link:ext:lna0) hogged as output/high
[    0.121028] GPIO line 19 (tp-link:ext:lna1) hogged as output/high
[    0.153584] workqueue: max_active 576 requested for napi_workq is out of range, clamping between 1 and 512
[    0.167791] clocksource: Switched to clocksource MIPS
[    0.174630] NET: Registered protocol family 2
[    0.179554] IP idents hash table entries: 2048 (order: 2, 16384 bytes, linear)
[    0.188065] tcp_listen_portaddr_hash hash table entries: 512 (order: 0, 4096 bytes, linear)
[    0.197047] TCP established hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.205233] TCP bind hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.212759] TCP: Hash tables configured (established 1024 bind 1024)
[    0.219681] UDP hash table entries: 256 (order: 0, 4096 bytes, linear)
[    0.226714] UDP-Lite hash table entries: 256 (order: 0, 4096 bytes, linear)
[    0.234523] NET: Registered protocol family 1
[    0.239254] PCI: CLS 0 bytes, default 32
[    0.247928] workingset: timestamp_bits=14 max_order=14 bucket_order=0
[    0.262933] squashfs: version 4.0 (2009/01/31) Phillip Lougher
[    0.269207] jffs2: version 2.2 (NAND) (SUMMARY) (LZMA) (RTIME) (CMODE_PRIORITY) (c) 2001-2006 Red Hat, Inc.
[    0.298493] pinctrl-single 1804002c.pinmux: 544 pins, size 68
[    0.305798] Serial: 8250/16550 driver, 1 ports, IRQ sharing disabled
[    0.313459] printk: console [ttyS0] disabled
[    0.318132] 18020000.uart: ttyS0 at MMIO 0x18020000 (irq = 9, base_baud = 2500000) is a 16550A
[    0.327351] printk: console [ttyS0] enabled
[    0.336259] printk: bootconsole [early0] disabled
[    0.365907] spi-nor spi0.0: s25fl064k (8192 Kbytes)
[    0.370938] 3 fixed-partitions partitions found on MTD device spi0.0
[    0.377400] Creating 3 MTD partitions on "spi0.0":
[    0.382283] 0x000000000000-0x000000020000 : "u-boot"
[    0.388570] 0x000000020000-0x0000007f0000 : "firmware"
[    0.397205] 2 tplink-fw partitions found on MTD device firmware
[    0.403277] Creating 2 MTD partitions on "firmware":
[    0.408354] 0x000000000000-0x0000001f989a : "kernel"
[    0.414499] 0x0000001f989c-0x0000007d0000 : "rootfs"
[    0.420642] mtd: device 3 (rootfs) set to be root filesystem
[    0.428569] 1 squashfs-split partitions found on MTD device rootfs
[    0.434923] 0x000000560000-0x0000007d0000 : "rootfs_data"
[    0.441512] 0x0000007f0000-0x000000800000 : "art"
[    0.449690] libphy: Fixed MDIO Bus: probed
[    0.792892] ag71xx 19000000.eth: Could not connect to PHY device. Deferring probe.
[    1.147784] random: fast init done
[    1.432213] libphy: ag71xx_mdio: probed
[    1.437446] libphy: ar8xxx-mdio: probed
[    1.449253] switch0: Atheros AR8229 rev. 1 switch registered on mdio.0
[    1.493522] ag71xx 1a000000.eth: connected to PHY at fixed-0:00 [uid=00000000, driver=Generic PHY]
[    1.503395] eth0: Atheros AG71xx at 0xba000000, irq 5, mode: gmii
[    1.509863] i2c /dev entries driver
[    1.516602] NET: Registered protocol family 10
[    1.527736] Segment Routing with IPv6
[    1.531663] NET: Registered protocol family 17
[    1.536292] 8021q: 802.1Q VLAN Support v1.8
[    1.857715] ag71xx 19000000.eth: connected to PHY at mdio.0:1f:00 [uid=004dd042, driver=Generic PHY]
[    1.868151] eth1: Atheros AG71xx at 0xb9000000, irq 4, mode: mii
[    1.880061] VFS: Mounted root (squashfs filesystem) readonly on device 31:3.
[    1.894228] Freeing unused kernel memory: 1184K
[    1.898861] This architecture does not have kernel memory protection.
[    1.905408] Run /sbin/init as init process
[    2.632932] init: Console is alive
[    2.636742] init: - watchdog -
[    3.857062] kmodloader: loading kernel modules from /etc/modules-boot.d/*
[    3.901864] kmodloader: done loading kernel modules from /etc/modules-boot.d/*
[    3.919758] init: - preinit -
[    5.467417] random: jshn: uninitialized urandom read (4 bytes read)
[    5.819155] random: jshn: uninitialized urandom read (4 bytes read)
[    6.039945] random: jshn: uninitialized urandom read (4 bytes read)
[    6.931652] eth0: link up (1000Mbps/Full duplex)
[    6.939888] IPv6: ADDRCONF(NETDEV_CHANGE): eth0: link becomes ready
[    6.955931] IPv6: ADDRCONF(NETDEV_CHANGE): eth0.1: link becomes ready
[    7.021000] urandom_read: 4 callbacks suppressed
[    7.021008] random: procd: uninitialized urandom read (4 bytes read)
[   11.236685] jffs2: notice: (513) jffs2_build_xattr_subsystem: complete building xattr subsystem, 10 of xdatum (0 unchecked, 2 orphan) and 12 of xref (2 dead, 0 orphan) found.
[   11.254123] mount_root: switching to jffs2 overlay
[   11.262761] overlayfs: upper fs does not support tmpfile.
[   11.274797] urandom-seed: Seeding with /etc/urandom.seed
[   11.389340] eth0: link down
[   11.418500] procd: - early -
[   11.421632] procd: - watchdog -
[   11.986216] procd: - watchdog -
[   11.991138] procd: - ubus -
[   12.073310] random: ubusd: uninitialized urandom read (4 bytes read)
[   12.082132] random: ubusd: uninitialized urandom read (4 bytes read)
[   12.093057] procd: - init -
[   13.311173] kmodloader: loading kernel modules from /etc/modules.d/*
[   13.382802] zram: Added device: zram0
[   13.427142] GACT probability on
[   13.437210] Mirror/redirect action on
[   13.475235] u32 classifier
[   13.478056]     input device check on
[   13.481781]     Actions configured
[   13.524660] Loading modules backported from Linux version v5.10.42-0-g65859eca4dff
[   13.532414] Backport generated by backports.git v5.10.42-1-0-gbee5c545
[   13.658021] xt_time: kernel timezone is -0000
[   13.696588] urngd: v1.0.2 started.
[   14.120794] PPP generic driver version 2.4.2
[   14.137031] NET: Registered protocol family 24
[   14.160718] random: crng init done
[   14.292303] ath: EEPROM regdomain sanitized
[   14.292315] ath: EEPROM regdomain: 0x64
[   14.292319] ath: EEPROM indicates we should expect a direct regpair map
[   14.292344] ath: Country alpha2 being used: 00
[   14.292348] ath: Regpair used: 0x64
[   14.306449] ieee80211 phy0: Selected rate control algorithm 'minstrel_ht'
[   14.308945] ieee80211 phy0: Atheros AR9340 Rev:2 mem=0xb8100000, irq=12
[   14.412314] kmodloader: done loading kernel modules from /etc/modules.d/*
[   18.506677] zram0: detected capacity change from 0 to 29360128
[   18.527047] Adding 28668k swap on /dev/zram0.  Priority:-2 extents:1 across:28668k SSDsc
[   32.131632] eth0: link up (1000Mbps/Full duplex)
[   32.146410] IPv6: ADDRCONF(NETDEV_CHANGE): eth0: link becomes ready
[   32.154343] br-lan: port 1(eth0.1) entered blocking state
[   32.159918] br-lan: port 1(eth0.1) entered disabled state
[   32.165918] device eth0.1 entered promiscuous mode
[   32.170845] device eth0 entered promiscuous mode
[   32.202403] br-lan: port 1(eth0.1) entered blocking state
[   32.207952] br-lan: port 1(eth0.1) entered forwarding state
[   33.145673] IPv6: ADDRCONF(NETDEV_CHANGE): br-lan: link becomes ready
[   34.361887] eth1: link up (100Mbps/Full duplex)
[   34.381127] IPv6: ADDRCONF(NETDEV_CHANGE): eth1: link becomes ready
[   35.101840] ath: EEPROM regdomain: 0x809e
[   35.101849] ath: EEPROM indicates we should expect a country code
[   35.101858] ath: doing EEPROM country->regdmn map search
[   35.101866] ath: country maps to regdmn code: 0x50
[   35.101874] ath: Country alpha2 being used: TW
[   35.101878] ath: Regpair used: 0x50
[   35.101886] ath: regdomain 0x809e dynamically updated by user
[   38.117922] br-lan: port 2(wlan0) entered blocking state
[   38.123388] br-lan: port 2(wlan0) entered disabled state
[   38.129192] device wlan0 entered promiscuous mode
[   38.215968] IPv6: ADDRCONF(NETDEV_CHANGE): wlan0: link becomes ready
[   38.222708] br-lan: port 2(wlan0) entered blocking state
[   38.228161] br-lan: port 2(wlan0) entered forwarding state
```

guipoletto commented:
is there a reason for merging this still being blocked?

hachazo commented:
I'm wondering the same thing, it's been a long time.

hauke commented:
Thank you for your patch, I applied it to master in 43ec6d64bb368253f159ec9e6d8343fdbb0d97e4.

jadermateus commented:
Hello, would you have the fullflash dump of the 841hp v2? I corrupted my bootloader and wanted to rewrite my flash

saif-ibne-ali commented:
> Hello, would you have the fullflash dump of the 841hp v2? I corrupted my bootloader and wanted to rewrite my flash

Sorry. I have no flash programmer tool. So, i cannot provide flash dump.
